Properly handle OpenAI JSON error responses when expecting stream response

galtonova opened the pull request:
OpenAI returns `application/json` content-type for errors even when expecting a stream in cases like token limit error or other openai related errors. Right now, `axios-fetch-adapter.js:getResponse` only throws a very generic error message, making it impossible to understand what the actual problem is.

```js
        // This is very generic and makes it impossible to understand the real error.
        throw new Error(
          `Expected content-type to be ${EventStreamContentType}, Actual: ${contentType}`
        );
```

My fix gets the json body and if there is a proper Error object found inside, throws the real error, making it possible to debug or notify users in such cases.

This issue also seems to be about the same problem: #670

nfcampos reviewed (commented):
This is a great improvement, but a small comment

galtonova commented on langchain/src/util/axios-fetch-adapter.js:
```diff
@@ -250,6 +250,16 @@ async function getResponse(request, config) {
+        if (contentType?.startsWith('application/json')) {
+          // If the response is JSON, try to parse it and throw a more specific error
+          response.data = await stageOne.json();
+          if (response.data?.error instanceof Error) {
```
In my tests I actually cared most about the second throw I added, with error.message. For this check, there were some fetch api implementations for node that would wrap 4xx and 5xx bodies into Error instances and I wanted to be on the safe side.

However, I checked the other parts of the code and I guess you only want to support native fetch api of Node 18. So you are right, that branch is completely unnecessary.

Now that I checked more about the codebase, you also wrap the returned response object of `getResponse` itself into `AxiosResponse` or `AxiosErrorResponse` object depending on the status, therefore even throwing anything is unnecessary. I can just return the response object here.

Also, from what I see, adding try/catch blocks around that `.json()` call is not needed because whatever `getResponse` throws is caught and thrown back to the handleLLMError callback (and also call method result).

I will edit that part now

galtonova commented:
I changed the code and made it directly return json or text error body. I still kept the old throw but now that thow only happens if the response status is 2xx and somehow non-stream content-type is received.

I also added some unit tests

nfcampos commented:
Thanks!
